### PR TITLE
`nixpkgs`: draw from `release-21.11` branch

### PR DESCRIPTION
**Pull request by ggreif**
# Porcupine

https://discourse.nixos.org/t/21-11-has-been-released/16353

---
Mostly benign changes, but we can now eliminate a bunch of custom pulls from Hackage, which is a bonus. (Actually this was possible with `21.05` too, but nobody noticed.)

## TODO (for follow-ups)
- eliminate/improve `nixpkgs-patched` (see comment below)
- transition to `llvm-13`

### CODE REVIEW

File: nix/drun.nix [hunk 1]
```diff
@@ -23,16 +23,16 @@ pkgs:
 
       buildInputs = with pkgs; [
         openssl
-        llvm_10
-        llvmPackages_10.libclang
+        llvm_12
```
Comment (ggreif) [inline nix/drun.nix:26]: As a follow-on PR I'll transition to `llvm_13`

File: nix/generate.nix [hunk 2]
```diff
@@ -53,9 +53,9 @@ let
     random = localHaskellSrc2nixWithDoc "qc-motoko" "test/random" "";
     lsp-int = localHaskellSrc2nixWithDoc "lsp-int" "test/lsp-int" "";
     # Packages on hackage that are newer than what's in nixpkgs
-    lsp-test = pkgs.haskellPackages.hackage2nix "lsp-test" "0.11.0.6";
-    haskell-lsp = pkgs.haskellPackages.hackage2nix "haskell-lsp" "0.23.0.0";
-    haskell-lsp-types = pkgs.haskellPackages.hackage2nix "haskell-lsp-types" "0.23.0.0";
+    #lsp-test = pkgs.haskellPackages.hackage2nix "lsp-test" "0.11.0.6";
```
Comment (ggreif) [inline nix/generate.nix:56]: Cite #2954 in a comment for what to do when a custom package should become necessary.

File: src/mo_frontend/parser.mly [hunk 3]
```diff
@@ -13,7 +13,7 @@ open Parser_lib
 
 let position_to_pos position =
   (* TBR: Remove assertion once the menhir bug is fixed. *)
-  assert (Obj.is_block (Obj.repr position));
+  assert Obj.(is_block (repr position));
```
Comment (ggreif) [inline src/mo_frontend/parser.mly:16]: Check this!

File: test/lsp-int/Main.hs [hunk 4]
```diff
@@ -72,7 +72,7 @@ definitionsTestCase
   -> [(FilePath, Range)]
   -> Session ()
 definitionsTestCase project doc pos expected = do
-  response <- getDefinitions doc pos
+  LSP.InL response <- getDefinitions doc pos
```
Comment (ggreif) [inline test/lsp-int/Main.hs:75]: this could `fail`, but we'll see when it happens!

File: nix/default.nix [hunk 5]
```diff
@@ -72,9 +71,11 @@ let
 
         # to allow picking up more recent Haskell packages from Hackage
         (self: super: {
-          all-cabal-hashes = self.fetchurl {
-            url = "https://github.com/commercialhaskell/all-cabal-hashes/archive/66a799608f2c6e0e6c530383bc0e2bcb42ae11f2.tar.gz";
-            sha256 = "0ds95gacrzsqg5f0f6j533ghxzcqqn7wn1d391pcpj5g9frp01q2";
+          all-cabal-hashes = self.fetchFromGitHub {
```
Comment (ggreif) [inline nix/default.nix:74]: Revert this!

File: nix/default.nix [hunk 6]
```diff
@@ -46,7 +45,7 @@ let
         # Rust nightly
         (self: super: let
           moz_overlay = import self.sources.nixpkgs-mozilla self super;
-          rust-channel = moz_overlay.rustChannelOf { date = "2021-10-25"; channel = "nightly"; };
+          rust-channel = moz_overlay.rustChannelOf { date = "2021-12-02"; channel = "nightly"; };
```
Comment (ggreif) [inline nix/default.nix:48]: A very happy palindrome! :-)

File: nix/default.nix [hunk 7]
```diff
@@ -13,9 +13,8 @@ let
   nixpkgs-patched = bootstrap-pkgs.applyPatches {
```
Comment (ggreif) [inline nix/default.nix:13]: @nomeata I tried to short-cut the `nixpkgs-patched` by eliminating, but it didn't work out. I didn't invest too much time, though...

File: test/lsp-int/Main.hs [hunk 8]
```diff
@@ -243,7 +243,7 @@ main = do
           -- ==> 1 | ort List
           let edit = TextEdit (Range (Position 0 1) (Position 0 3)) ""
           _ <- applyEdit doc edit
-          sendNotification TextDocumentDidSave (DidSaveTextDocumentParams doc)
+          sendNotification STextDocumentDidSave (DidSaveTextDocumentParams doc Nothing)
```
Comment (ggreif) [inline test/lsp-int/Main.hs:246]: maybe there is a helper for this?

File: nix/generate.nix (hunk 2 above)
Comment (ggreif) [reply nix/generate.nix]: Done: 69d49ee.

File: nix/default.nix (hunk 5 above)
Comment (ggreif) [reply nix/default.nix]: Done!

File: test/lsp-int/Main.hs (hunk 8 above)
Comment (ggreif) [reply test/lsp-int/Main.hs:246]: Looked in https://github.com/haskell/lsp.git today, but nothing obvious...

File: src/mo_frontend/parser.mly (hunk 3 above)
Comment (ggreif) [reply src/mo_frontend/parser.mly]: Tentatively removed in bde3a44.

File: nix/default.nix (hunk 7 above)
Comment (nomeata) [reply nix/default.nix:13]: You could copy what I did at https://github.com/dfinity/ic-hs/blob/master/nix/default.nix#L9, but it's not important

**Review by nomeata — APPROVED** [review]
Good job!

**Comment by ggreif** [comment]
@Mergifyio refresh

**Pull request merged**